### DISCUSSION TITLE
FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg'

### DISCUSSION THREAD

**Discussion by JeanVel** [openai/whisper#1379] [Show and tell]
Hello, I am using a Mac with a M1 chip and I have an issue. 
I tried to install whisper to transcribe an audio, I followed the instruction, I managed to install the package whisper, and ffmpeg (using homebrew). But when I use the following command : transcription = model.transcribe(audio_path, fp16=False), I get the following error : FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg'. Does anyone has an idea of how I could fix this ? Thank you

**Comment by yasersakkaf** [comment]
Facing the same issue, any solutions yet?

**Reply by SAYUK09** [reply]
Encountered this on ubuntu system and the following solved it for me.

Step 1:
sudo apt update && sudo apt install ffmpeg

Step 2:
sudo apt install ffmpeg

Hope this helps.

**Comment by Rentetier** [comment]
I have the same issue, any solutions?

**Reply by SAYUK09** [reply]
This helped me to solve it on ubuntu
Just encountered the same error on ubuntu, and solved it using the following way.

Step 1:
sudo apt update && sudo apt install ffmpeg

Step 2:
sudo apt install ffmpeg

**Comment by glangford** [comment]
`ffmpeg` must be in your PATH for `whisper` to use it. In terminal try `which ffmpeg` to verify. Otherwise, see here for installation  options:

* https://superuser.com/questions/624561/install-ffmpeg-on-os-x

Or this prior discussion may be helpful:
* https://github.com/openai/whisper/discussions/1238

**Reply by guus-vrv** [reply]
This worked for me, thanks!

**Reply by ghost** [reply]
This worked for me, too. Thank you !

**Comment by minerb007** [comment]
once I installed FFMPEG and add it to the path, I still had issues running it via terminal in VS Code, but running it through windows command prompt solved my issues

**Reply by Codie-Petersen** [reply]
This user's problem is a bit different than yours. They are doing it for Mac. Sometimes you have to reload VSCode to get the path to load correctly. Definitely make sure you reload the specific terminal you were using in VSCode because it won't know anything about the updated Path unless it was started after Path was updated.

For any other windows users here are the install instructions for FFMPEG:
https://phoenixnap.com/kb/ffmpeg-windows

**Comment by paranoid2droid** [comment]
In my case, ffmpeg was installed in the base environment. When I switched to another conda environment, the same error appears. If you are also using conda, the simplest solution might be just to install ffmpeg via conda:
`conda install -c conda-forge ffmpeg`

**Comment by SAYUK09** [comment]
Just encountered the same error on ubuntu, and solved it using the following way.

- Step 1:
`sudo apt update && sudo apt install ffmpeg`

- Step 2:
`sudo apt install ffmpeg`

Hope this helps.

**Reply by zh1109334794** [reply]
This worked for me,thanks

**Comment by Yuanyuan-888** [comment]
Hi, everyone,

I am quite confusing with this error.

As long as I am using a40 gpu in my university HPC to test with whisper-large model, it shows this ffmpeg error! But when I used v100 gpu, this error is gone, and the whisper-large can successfully decode something.

When I use "which ffmpeg", it shows the location "/usr/bin/ffmpeg"
and in this conda envirnment, when I "import ffmpeg", it is also successful.

By the way, on our hpc system, we do not have sudo accessment TT

If anyone has some idea about why this error is gpu dependent, please do leave me a message, need you help!

**Comment by okineadev** [comment]
Can you provide logs?

**Reply by Yuanyuan-888** [reply]
2024-04-27T08:27:44 (evaluate_asr.sh:106:main) scripts/utils/evaluate_asr.sh --whisper_tag large --nj 2 --gpu_inference true --stage 2 --stop_stage 3 --cleaner whisper_basic --hyp_cleaner whisper_basic --decode_options {language: nl, task: transcribe, temperature: 0, beam_size: 10, fp16: False} --gt_text dump/raw/dev_s/text dump/raw/dev_s/wav.scp whisper-large_outputs_nl_0424/dev_s
2024-04-27T08:27:50 (evaluate_asr.sh:160:main) stage 2: ASR inference
2024-04-27T08:27:50 (evaluate_asr.sh:191:main) Decoding started... log: 'whisper-large_outputs_nl_0424/dev_s/logdir/asr_inference.*.log'
run.pl: job failed, log is in whisper-large_outputs_nl_0424/dev_s/logdir/asr_inference.1.log
# 
python3 pyscripts/utils/evaluate_whisper_inference.py --ngpu 1 --data_path_and_name_and_type dump/raw/dev_s/wav.scp --key_file whisper-large_outputs_nl_0424/dev_s/logdir/keys.1.scp --model_tag large --model_dir whisper-large_outputs_nl_0424/dev_s/models --output_dir whisper-large_outputs_nl_0424/dev_s/logdir/output.1 --decode_options "{language: nl, task: transcribe, temperature: 0, beam_size: 10, fp16: False}"
# 
Started at Sat Apr 27 08:27:50 CEST 2024
# 
/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/site-packages/whisper/timing.py:57: NumbaDeprecationWarning: ^[[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.^[[0m
  def backtrace(trace: np.ndarray):
/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/bin/python3 pyscripts/utils/evaluate_whisper_inference.py --ngpu 1 --data_path_and_name_and_type dump/raw/dev_s/wav.scp --key_file whisper-large_outputs_nl_0424/dev_s/logdir/keys.1.scp --model_tag large --model_dir whisper-large_outputs_nl_0424/dev_s/models --output_dir whisper-large_outputs_nl_0424/dev_s/logdir/output.1 --decode_options '{language: nl, task: transcribe, temperature: 0, beam_size: 10, fp16: False}'
Traceback (most recent call last):
  File "pyscripts/utils/evaluate_whisper_inference.py", line 195, in <module>
    main()
  File "pyscripts/utils/evaluate_whisper_inference.py", line 191, in main
    inference(**kwargs)
  File "pyscripts/utils/evaluate_whisper_inference.py", line 103, in inference
    results = speech2text(os.path.abspath(audio_file.strip()), **decode_options)
  File "/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "pyscripts/utils/evaluate_whisper_inference.py", line 50, in __call__
    result = self.model.transcribe(speech, **decode_options)
  File "/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/site-packages/whisper/transcribe.py", line 121, in transcribe
    mel = log_mel_spectrogram(audio, padding=N_SAMPLES)
  File "/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/site-packages/whisper/audio.py", line 130, in log_mel_spectrogram
    audio = load_audio(audio)
  File "/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/site-packages/whisper/audio.py", line 46, in load_audio
    ffmpeg.input(file, threads=0)
  File "/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/site-packages/ffmpeg/_run.py", line 313, in run
    process = run_async(
  File "/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/site-packages/ffmpeg/_run.py", line 284, in run_async
    return subprocess.Popen(
  File "/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/tudelft.net/staff-bulk/ewi/insy/SpeechLab/yuanyuanzhang/espnet_yyy/espnet/tools/miniconda/envs/espnet/lib/python3.8/subprocess.py", line 1720, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg'
# 
Accounting: time=107 threads=1
# 
Ended (code 1) at Sat Apr 27 08:29:37 CEST 2024, elapsed time 107 seconds
srun: error: gpu16: task 0: Exited with exit code 1
srun: launch/slurm: _step_signal: Terminating StepId=9923822.0

This is the error infor on a a40 gpu. but when I decode with the same code on v100, then no error

**Reply by Yuanyuan-888** [reply]
Hi, I figured out, I think it fails because on my gpu nodes, it did not have ffmpeg command line interface.

**Comment by Yuanyuan-888** [comment]
Hi,  I still have a question, when directly use whisper-large model to decode, in the code miniconda/envs/espnet/lib/python3.8/site-packages/whisper/audio.py

it uses ffmpeg to load the audio but not torchaudio to load the audio.

def load_audio(file: str, sr: int = SAMPLE_RATE):
    """ 
    Open an audio file and read as mono waveform, resampling as necessary

    Parameters
    ----------
    file: str
        The audio file to open

    sr: int
        The sample rate to resample the audio if necessary

    Returns
    -------
    A NumPy array containing the audio waveform, in float32 dtype.
    """
    try:
        # This launches a subprocess to decode audio while down-mixing and resampling as necessary.
        # Requires the ffmpeg CLI and `ffmpeg-python` package to be installed.
        out, _ = (
            ffmpeg.input(file, threads=0)
            .output("-", format="s16le", acodec="pcm_s16le", ac=1, ar=sr)
            .run(cmd=["ffmpeg", "-nostdin"], capture_stdout=True, capture_stderr=True)
        )
    except ffmpeg.Error as e:
        raise RuntimeError(f"Failed to load audio: {e.stderr.decode()}") from e

    return np.frombuffer(out, np.int16).flatten().astype(np.float32) / 32768.0

If I use torchaudio to replace ffmpeg,

import torchaudio
import torch

def load_audio(file: str, sr: int = SAMPLE_RATE):
    """ 
    Open an audio file and read as mono waveform, resampling as necessary

    Parameters
    ----------
    file: str
        The audio file to open

    sr: int
        The sample rate to resample the audio if necessary

    Returns
    -------
    A NumPy array containing the audio waveform, in float32 dtype.
    """
    try:
        # Load audio using torchaudio
        waveform, _ = torchaudio.load(file, normalize=True, channels_first=True, num_frames=-1)
        
        # Resample if necessary
        if sr != SAMPLE_RATE:
            waveform = torchaudio.transforms.Resample(orig_freq=SAMPLE_RATE, new_freq=sr)(waveform)
        
    except Exception as e:
        raise RuntimeError(f"Failed to load audio: {e}") from e

    # Convert to NumPy array and return
    return waveform.numpy().astype(np.float32)

I think these are the same.
